packages/multi-test/src/app.rs

```diff
@@ -19,153 +19,345 @@ use crate::wasm::{ContractData, Wasm, WasmKeeper};
-    block: Option<BlockInfo>,
+pub struct AppBuilder<Bank, Api, Storage, Custom, Wasm> {
+    wasm: Wasm,
+    bank: Bank,
+    api: Api,
+    storage: Storage,
+    custom: Custom,
+    block: BlockInfo,
 }
 
-impl<ExecC, QueryC> AppBuilder<ExecC, QueryC>
+impl Default
+    for AppBuilder<
+        BankKeeper,
+        MockApi,
+        MockStorage,
+        PanickingCustomHandler<Empty, Empty>,
+        WasmKeeper<Empty, Empty>,
+    >
+{
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl
+    AppBuilder<
+        BankKeeper,
+        MockApi,
+        MockStorage,
+        PanickingCustomHandler<Empty, Empty>,
+        WasmKeeper<Empty, Empty>,
+    >
+{
+    /// Creates builder with default components working with empty exec and query messages.
+    pub fn new() -> Self {
+        AppBuilder {
+            wasm: WasmKeeper::new(),
+            bank: BankKeeper::new(),
+            api: MockApi::default(),
+            storage: MockStorage::new(),
+            custom: PanickingCustomHandler::new(),
+            block: mock_env().block,
+        }
+    }
+}
+
+impl<ExecC, QueryC>
+    AppBuilder<
+        BankKeeper,
+        MockApi,
+        MockStorage,
+        PanickingCustomHandler<ExecC, QueryC>,
+        WasmKeeper<ExecC, QueryC>,
+    >
 where
-    ExecC: Debug + PartialEq + Clone + JsonSchema + 'static,
+    ExecC: Debug + Clone + PartialEq + JsonSchema + DeserializeOwned + 'static,
     QueryC: Debug + CustomQuery + DeserializeOwned + 'static,
 {
-    /// Overwrites default wasm executor. Panic if already set.
-    #[track_caller]
-    pub fn with_wasm(mut self, wasm: impl Wasm<ExecC, QueryC> + 'static) -> Self {
-        assert!(self.wasm.is_none(), "Wasm executor already overwritten");
-        self.wasm = Some(Box::new(wasm));
-        self
+    /// Creates builder with default components designed to work with custom exec and query
+    /// messages.
+    pub fn new_custom() -> Self {
+        AppBuilder {
+            wasm: WasmKeeper::new(),
+            bank: BankKeeper::new(),
+            api: MockApi::default(),
+            storage: MockStorage::new(),
+            custom: PanickingCustomHandler::new(),
+            block: mock_env().block,
+        }
+    }
+}
+
+impl<BankT, ApiT, StorageT, CustomT, WasmT> AppBuilder<BankT, ApiT, StorageT, CustomT, WasmT> {
+    /// Overwrites default wasm executor.
+    ///
+    /// At this point it is needed that new wasm implements some `Wasm` trait, but it doesn't need
+    /// to be bound to Bank or Custom yet - as those may change. The cross-components validation is
+    /// done on final building.
+    ///
+    /// Also it is possible to completely abandon trait bounding here which would not be bad idea,
+    /// however it might make the message on build creepy in many cases, so as for properly build
+    /// `App` we always want `Wasm` to be `Wasm`, some checks are done early.
+    pub fn with_wasm<B, C: CustomHandler, NewWasm: Wasm<C::ExecC, C::QueryC>>(
```

Comment:
Huh.
I don't like so much disconnecting the Wasm<C::ExecC, C::QueryC> from they types used in CustomT (which is already defined).

So, you need to specify the proper (final) ExecC/QueryC when creating the (empty) builder, then can only install a new WasmKeeper that uses the same custom types.

That doesn't seem to be enforced here. But maybe I am just not following this.

Reply:
Yes, you are wrong. It took me the most time to go around this. The only solution I found which actually connects wasm with other types requires, and plays well with building app is requiring GATs, so there are to reasons I didn't went for it:
* It requires GATs so it is an opposite of being straightforward
* It requires GATs, and GATs are not yet in stable.

I would love to find good solution for properly connecting those two. And I am sure it is one. I think the problem is, that those types are so entangled that gradually simplifying bounds could help, but I don't see the solution yet.

Reply:
Basically the problem is, that you have to disconnect those types here. If you wont, then how would you even build the app for non-standard custom msgs? If you set custom handler first, it fails - your new custom handler types doesn't match wasm types. Same if you set wasm first, the again, you are in the same place (unless wasm can handle both types which is true for probably most cases, but in some edge it may not - at the and wasm may be substituted with custom one). The relation between those types is defined in build, as then both components have to be final.

Reply:
Let's discuss this in a future issue. It is good (and possibly complete correct) as it is. Maybe I try a small PR later on to address this, but definitely not a blocker to merge.